### ISSUE TITLE
Unable to send email due to double EHLO command

### ISSUE THREAD

**Issue by mgee** [hoaproject/Mail#21]
When sending an email via `somehost.com`  (replaced the original hostname) using `STARTTLS` and `AUTH LOGIN` the following exception is thrown:

```
[2015-06-10T17:06:25+00:00] "Authentification failed (LOGIN) (code: 250, message: “somehost.com”)." vendor/hoa/mail/Transport/Smtp.php:188
```

In `Transport/Smtp.php` `EHLO` is send to the server twice. Once in https://github.com/hoaproject/Mail/blob/16b512403f75551a9990996bdbe8f0b05c895e86/Transport/Smtp.php#L216
and a second time in
https://github.com/hoaproject/Mail/blob/16b512403f75551a9990996bdbe8f0b05c895e86/Transport/Smtp.php#L231

Removing the second `EHLO` fixes the error.

**Comment by Hywan** [comment]
Hello :-),

When we change the connection encryption (here with `STARTTLS`), we have to re-send a `EHLO` because the content can change. This is the short-version. The long-version is detailed in Section 4.2 of the [RFC3207](https://tools.ietf.org/html/rfc3207):

> 4.2 Result of the STARTTLS Command
>
> Upon completion of the TLS handshake, the SMTP protocol is reset to
> the initial state (the state in SMTP after a server issues a 220
> service ready greeting).  The server MUST discard any knowledge
> obtained from the client, such as the argument to the EHLO command,
> which was not obtained from the TLS negotiation itself.  The client
> MUST discard any knowledge obtained from the server, such as the list
> of SMTP service extensions, which was not obtained from the TLS
> negotiation itself.  The client SHOULD send an EHLO command as the
> first command after a successful TLS negotiation.
>
> The list of SMTP service extensions returned in response to an EHLO
> command received after the TLS handshake MAY be different than the
> list returned before the TLS handshake.  For example, an SMTP server
> might not want to advertise support for a particular SASL mechanism
> [SASL] unless a client has sent an appropriate client certificate
> during a TLS handshake.
>
> Both the client and the server MUST know if there is a TLS session
> active.  A client MUST NOT attempt to start a TLS session if a TLS
> session is already active.  A server MUST NOT return the STARTTLS
> extension in response to an EHLO command received after a TLS
> handshake has completed.

So this is correct and I don't see how it can be bad. Even sending `EHLO` twice immediatly would not alter the workflow.

Your error is about the authentication though. Actually, this is this block code: https://github.com/hoaproject/Mail/blob/16b512403f75551a9990996bdbe8f0b05c895e86/Transport/Smtp.php#L260-L270. It is not about an incorrect username/password pair. The SMTP client sent a `LOGIN` command because `somehost.com` allows the `AUTH LOGIN` authentication. However, when the SMTP client sends this command, `somehost.com` says it's an error. I would reckon there is an error on your server but I am very curious. Can you dump the responses from the server by basically inserting a `var_dump($ehlo);` right here https://github.com/hoaproject/Mail/blob/a3847f746cefbf93090ce32e4655a2dd8dca5328/Transport/Smtp.php#L234?

**Comment by Hywan** [comment]
I just sent PR #24, which is not totally related to your issue though: It avoids sending a second `EHLO` in case `STARTTLS` is not supported by the server. This is different of your problem since you have `STARTTLS` is supported by your server.

**Comment by Hywan** [comment]
Another question: Are you on Windows? It would give me another clue.

**Comment by mgee** [comment]
Here is the content of the `$ehlo` variable:
```
Array
(
[0] => 250-STARTTLS
[1] => PIPELINING
[2] => 8BITMIME
[3] => AUTH LOGIN
)
```

The server is running Linux with netqmail.

**Comment by Hywan** [comment]
Ok thanks, I will try to reproduce the bug. Maybe @Savageman, @Pierozi or @sekiltoyai could also help!

**Comment by Pierozi** [comment]
Hello @mgee, 

what version of PHP do you use ?
Can you please try this `telnet` command for be sure Server are correctly configured ?

```bash
$ user@domain : telnet smtp.domain.tld 587
Trying 7.7.7.7...
Connected to smtp.domain.tld.
Escape character is '^]'.
220 mail.domain.tld ESMTP Postfix (Debian/GNU)
HELO <address@domain.tld>
250 mail.domain.tld
STARTTLS
220 2.0.0 Ready to start TLS
```

**Comment by mgee** [comment]
PHP version 5.6.6

Here is the output of the telnet command:
```
$ telnet smtp.domain.tld 587
Connected to smtp.domain.tld.
Escape character is '^]'.
220 smtp.domain.tld ESMTP
HELO smtp.domain.tld
250 smtp.domain.tld
STARTTLS
220 Proceed.
```

**Comment by Pierozi** [comment]
I've done some test in `5.5.9` and `5.6.9` and it's fine.
@mgee  can you try with that Snippet below and with fresh composer require.

```php
<?php

require 'vendor/autoload.php';

Hoa\Mail\Message::setDefaultTransport(
    new Hoa\Mail\Transport\Smtp(
        new Hoa\Socket\Client('tcp://domain.tld:587'),
        'REAL_SENDER@domain.tld', //user
        'PASSWORD'
    )
);

$message            = new Hoa\Mail\Message();
$message['From']    = 'Gordon Freeman <REAL_SENDER@domain.tld>';
$message['To']      = 'Your Name <REAL_RECEIVER@domain.tld>';
$message['Subject'] = 'Hoa is awesome!';

$message->addContent(
    new Hoa\Mail\Content\Text('Check this out: http://hoa-project.net/!')
);

$message->send();
```

**Comment by mgee** [comment]
hoa/mail `0.15.06.03`

PHP Fatal error:  Uncaught Hoa\Mail\Transport\Smtp::ifNot(): (0) Authentification failed (LOGIN) (code: 250, message: “domain.tld”).

**Comment by mgee** [comment]
Using hoa/mail dev-master I get the following error:
```
PHP Fatal error:  Uncaught Hoa\Mail\Transport\Smtp::send(): (2) The server does not support authentication, we cannot authenticate.
in /home/mgee/test/vendor/hoa/mail/Transport/Smtp.php at line 243.
  thrown in /home/mgee/test/vendor/hoa/mail/Transport/Smtp.php on line 243
```

Inserting a `var_dump($ehlo);` here:
https://github.com/hoaproject/Mail/blob/8d5e144d5438aafe9b7b87026990cfbc263e4788/Transport/Smtp.php#L241

Yields this:
```
array(1) {
  [0]=>
  string(26) "250-domain.tld
"
}
```

**Comment by Hywan** [comment]
So on `hoa/mail@dev`, the `$ehlo` variable does contain only `250-domain.tld`, nothing else. I assume it sends the `STARTTLS` command before. My guess is that the server is not well-configured. Sending a `EHLO` before and after the `STARTTLS` might not produce the same result (see the quoted RFC above) because some commands would not be accessible the same way. Did you install the SMTP server by yourself?

I am surprised though that there is a difference between `hoa/mail@0.15.06.03` and `hoa/mail@dev`. The workflow is the same. Could you please re-run a `telnet` but with `EHLO` instead of `HELO`, something like this:

```
$ telnet smtp.domain.tld 587
Connected to smtp.domain.tld.
Escape character is '^]'.
220 smtp.domain.tld ESMTP
EHLO smtp.domain.tld
…
```

You should see `STARTTLS` here but maybe the `AUTH` could appear before. That's just limited that we cannot send any command after that…

**Comment by Hywan** [comment]
In fact, I am surprised that your server tells the client it is possible to authenticate with `LOGIN` while when we send the `AUTH LOGIN` command, it fails. It's like: “I can do that. — Ok do it. — No I can't.”.

**Comment by mgee** [comment]
It seems that the `$client->read(2048);` after the `EHLO` command only reads the first line of the server response. This is `250-domain.tld`.

Directly calling `$client->read(2048);` again reads the remaining data:
```
array(4) {
  [0]=>
  string(12) "250-STARTTLS"
  [1]=>
  string(10) "PIPELINING"
  [2]=>
  string(8) "8BITMIME"
  [3]=>
  string(18) "AUTH LOGIN PLAIN
"
}
```

Inserting a `sleep(1);` after sending the first `EHLO` fixes this. Then the first read already contains all data. 
However, using the same trick for the second `EHLO` in the `STARTTLS` case does not work.

The SMTP server is a host from https://uberspace.de (you can get a free trial account there).

**Comment by Hywan** [comment]
What if you change for `$client->read(4096);`?

**Comment by Hywan** [comment]
(I won't be present the next days, I am assigning @Pierozi, sorry, good reasons :-)).

**Comment by mgee** [comment]
Changing the read to `$client->read(4096);` has no effect. I suspect it is more of a timing issue rather than the length of the read buffer.

**Comment by Pierozi** [comment]
@mgee thanks for sharing the host, i will create an account on uberspace and find issue.
Back to you then.

**Comment by Pierozi** [comment]
I've found the reason of the BUG.

Hoa use `read` method for retrieve data once something come.
But, like is the steam, data can come by piece. 
And like the Server has some latency we only get the first part `250-shaula.uberspace.de`

```php
    $client->writeAll('EHLO ' . $domain . CRLF);

  sleep(1);

        $ehlo = preg_split('#' . CRLF . '250[\-\s]+#', $client->read(2048));

    var_dump($ehlo);
/**
array(5) {
  [0]=>
  string(23) "250-shaula.uberspace.de"
  [1]=>
  string(8) "STARTTLS"
  [2]=>
  string(10) "PIPELINING"
  [3]=>
  string(8) "8BITMIME"
  [4]=>
  string(18) "AUTH LOGIN PLAIN
"
}*/
```

If I put `sleep` one second between write & read. the $ehlo contain all parts missing.
But it's fail again at second EHLO before AUTH.

**Comment by Hywan** [comment]
Ok, that's what I was suspecting. I see. Hmm, we should see if something is still on the buffer, and if yes, try a sleep. Maybe @Pierozi has another better idea?

**Comment by Hywan** [comment]
We should add a `setStreamReadBuffer` method on `Hoa\Stream`. @Pierozi, could you try with `stream_set_read_buffer($client->getStream(), 1024);` (maybe it could solve the issue). I don't know what is the default size of the read buffer. Any idea?

**Comment by Pierozi** [comment]
@Hywan it's `STREAM PEEK` or `OUT OF BAND` mode ?
i will test with method `stream_set_read_buffer`

**Comment by Sekiltoyai** [comment]
Hello @Hywan and @Pierozi,
The right method for testing incoming data is `select()`. Each time you don't know exactly what will be the received length, you need to call `select()` until your message is complete. In your case, loop on `select()` until you reach a CRLF.
By the way, you may also need to bufferize data in the SMTP class, in case the server sends multiple lines in a row, making you to have multiple lines in your buffer.

**Comment by Hywan** [comment]
@Pierozi Yup, exact.
@Sekiltoyai I didn't understand your proposal about bufferizing data in the SMTP class. Could you provide an example please?

**Comment by Sekiltoyai** [comment]
@Hywan :
`$line_completed = false;
$buffer = "";
$r = $socket;
while(!$line_completed) {
    if (stream_select($r, NULL, NULL 1)) {
        $buffer .= stream_socket_recvfrom($socket, 1024);
        if ( /* test if line is complete */ )
            $line_completed = true;
    }
}`
Untested, but the equivalent would definitely work in C.

**Comment by Hywan** [comment]
@Sekiltoyai About the idea of `STREAM PEEK` or `OUT OF BAND`? I guess that's better, no?

**Comment by Sekiltoyai** [comment]
STREAM_OOB has nothing to do with your problem. It just allows to read data that have been sent as out-of-band, e.g. in TCP with the URG flag set.
STREAM_PEEK may be useful if you don't want to consume the buffer when you read data, as explained in the documentation. In that case you may read data by picking it until you reach the newline, then read normally to effectively consume the buffer. So it would exclude the need for an additional buffer in the application.
Is your read operation blocking ?

**Comment by Hywan** [comment]
@Sekiltoyai By default it depends of the OS. We can force to be blocking though. I think it will solve our issue. Do you agree?

**Comment by Hywan** [comment]
I guess we have to block yes. We don't wait on multiple sockets and PHP is not asynchronous so it's fine here :wink:.

**Comment by Hywan** [comment]
Sorry, Waffle changed the assignee.

**Comment by Hywan** [comment]
@Pierozi Could you test https://github.com/hoaproject/Mail/pull/25 please? I guess it should fix this issue. I have no time to test but maybe you have a specific script to test the timeout.

**Comment by Hywan** [comment]
@mgee Please, could you test with the patch in #25 to see if it fixes your issue?

**Comment by mgee** [comment]
@Hywan unfortunately patch #25 does not fix my issue. The `ehlo` variable is still:
```
array(1) {
  [0]=>
  string(26) "250-domain.tld
"
}
```

**Comment by Hywan** [comment]
@mgee Damned… How can I reproduce it easily?

**Comment by mgee** [comment]
@Hywan the easiest way seems to be to sign up for a trial account on https://uberspace.de. They require almost no data for the signup and offer ssh access.

**Comment by Hywan** [comment]
@mgee Good. Thanks!

**Comment by Hywan** [comment]
@mgee I have created a new account. What should I do to activate my email address?

**Comment by mgee** [comment]
@Hywan your email address is displayed in bold on the first paragraph of this [page](https://uberspace.de/dashboard/mail) and does not need to be activated. To be able to use the ssh access to your uberspace set a password (or ssh key) on this [page](https://uberspace.de/dashboard/authentication). Then you can login via ssh and install composer etc.

**Comment by Hywan** [comment]
@mgee What you are trying to achieve is sending an email with their SMTP server, is that correct? If yes, could you provide a minimal working example please?

**Comment by mgee** [comment]
@Hywan yes. Here is the example code I am using. You just have to change the username, password and email addresses.

```php
<?php

require 'vendor/autoload.php';

Hoa\Mail\Message::setDefaultTransport(
    new Hoa\Mail\Transport\Smtp(
        new Hoa\Socket\Client('tcp://localhost:587'),
        'username', //user
        'password'
    )
);

$message            = new Hoa\Mail\Message();
$message['From']    = 'Gordon Freeman <email@domain.de>';
$message['To']      = 'Your Name <email@domain.de>';
$message['Subject'] = 'Hoa is awesome!';

$message->addContent(
    new Hoa\Mail\Content\Text('Check this out: http://hoa-project.net/!')
);

$message->send();
```

**Comment by Hywan** [comment]
@mgee What if you don't use `localhost` but `userspace.de` public URI instead?

**Comment by mgee** [comment]
Same error, the `ehlo` variable is still incomplete.

**Comment by Sekiltoyai** [comment]
@Hywan 
Blocking is useful but not sufficient, you still have to read() until you reach a newline in the tcp stream.

**Comment by Hywan** [comment]
@Sekiltoyai Yup. Thanks!

**Comment by Hywan** [comment]
@mgee Still trying to find the best solution with @Sekiltoyai (by emails though). We do not forget you!

**Comment by Hywan** [comment]
@mgee I guess it works now. Making another PR.

**Comment by Hywan** [comment]
@mgee I am pretty sure https://github.com/hoaproject/Mail/pull/27 will solve your issue. I introduce timeout support and a better line reading, see the PR description. /cc @Sekiltoyai 

I will merge as soon as I receive your feedback. My tests are green.

**Comment by mgee** [comment]
@Hywan  #27 fixes the issue. Thanks! :+1:

**Comment by Hywan** [comment]
@mgee A new snapshot has been released because this is a bug fix: https://github.com/hoaproject/Mail/releases/tag/0.15.08.28.
